x-pack/metricbeat/module/oracle/tablespace/metricset.go

```diff
@@ -0,0 +1,77 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package tablespace
+
+import (
+	"database/sql"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/oracle"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("oracle", "tablespace", New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	db        *sql.DB
+	extractor tablespaceExtractMethods
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The oracle 'tablespace' metricset is beta.")
+
+	config := oracle.ConnectionDetails{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, errors.Wrap(err, "error parsing config file")
+	}
+
+	db, err := oracle.NewConnection(&config)
```

Comment:
we normally do this on fetch, as returning an error here will make Metricbeat stop, maybe the database is not accepting new connections now but goes online later. See this for example: https://github.com/elastic/beats/blob/master/metricbeat/module/mysql/status/status.go#L55-L61

Reply:
So what about this? https://github.com/elastic/beats/pull/12603

Reply:
That change adds a wrapper to the DB to reuse connectiongs, that's ok. If you look closely, initialization does nothing: https://github.com/elastic/beats/blob/5cd611aa84430f245c4cfb316eb6aab5e018c84a/metricbeat/module/postgresql/metricset.go#L40 to make sure connection errors don't happen there, but later on `Fetch`, when we call `DB`: https://github.com/elastic/beats/blob/5cd611aa84430f245c4cfb316eb6aab5e018c84a/metricbeat/module/postgresql/metricset.go#L45